src/frontend/src/optimizer/logical_optimization.rs

```diff
@@ -141,7 +141,12 @@ static GENERAL_UNNESTING_TRANS_APPLY_WITH_SHARE: LazyLock<OptimizationStage> =
     LazyLock::new(|| {
         OptimizationStage::new(
             "General Unnesting(Translate Apply)",
-            vec![TranslateApplyRule::create(true)],
+            vec![
+                TranslateApplyRule::create(true),
+                // Separate the project from a join if necessary because `ApplyJoinTransposeRule`
+                // can't handle a join with `output_indices`.
+                ProjectJoinSeparateRule::create(),
```

Comment:
Does it always cancel out with the `ProjectJoinMergeRule`? Did not quite get how the "if necessary" part works.

Reply:
Yes, it always cancels out with the `ProjectJoinMergeRule` unless the project of the join is a trivial project.